### PR TITLE
Update object conversion api

### PR DESCRIPTION
**Pull request by ColCarroll**
I would like to update docstrings a bit more, but I think this is ready to go, code-wise. The changes are described in detail [in this gist](https://gist.github.com/ColCarroll/369a3ba54b27b197b9b3e29289347ba5), but roughly this is just a refactor of everyone else's work to reuse some patterns that were coming up.

The end result is an api that is much more flexible (see screenshot for plotting dictionaries, numpy arrays, and netcdf files from disk). This also adds more flexible utilities for `pystan` and `pymc3`, and adds support for handling posterior predictive and prior data from PyMC3.

This removes the reliance on having a file as a backend. It probably would not be too hard to allow that, but one has to be careful with opening and closing the netCDF file.

image

### CODE REVIEW

**Comment by aloctavodia** [comment]
This is fantastic! 

We should rename `violintraceplot` to just `violinplot`, when I name it I was thinking on one function for "general data" and another for traces, but this is much better.

**Comment by canyon289** [comment]
I also agree this is fantastic. I see no blockers for a merge.

I do have a lot of comments and questions though that I'll pepper all over the place. The biggest overarching comment I have though is something you already mentioned, I think there needs to be way more documentation. I'm particularly motivated to mention this given that when I tried working on pymc3 it was really difficult to understand why code paths existed, or what cases they were handling.

**Comment by ahartikainen** [comment]
Could rhat plot be semilog? I mean when it fails, it can be huge.

**Comment by ahartikainen** [comment]
I mean linear near 1.

File: arviz/utils/xarray_utils.py [hunk 1]
```diff
@@ -1,454 +1,246 @@
...
+            can_squeeze = True # added a dimension, might remove it too
+        n_chains, n_samples, *shape = ary.shape
+        if n_chains > n_samples:
+            warnings.warn('More chains ({n_chains}) than draws ({n_samples}). '
+                          'Passed array should have shape (chains, draws, *shape)'.format(
+                              n_chains=n_chains, n_samples=n_samples), SyntaxWarning)
+
+        coords = {}
+        dims = default_dims
+        if can_squeeze and len(shape) == 1 and shape[0] == 1:
```
Comment (canyon289) [inline arviz/utils/xarray_utils.py]: Perhaps a comment here on what case this handles would be good. It's not obvious, at least to me what this checks for and what problem it solves

File: arviz/utils/xarray_utils.py [hunk 2]
```diff
@@ -1,454 +1,246 @@
...
-        if coords is None:
-            coords = {}
-
-        coords['chain'] = np.arange(chains)
-        super().__init__(obj, filename=filename, coords=coords, dims=dims)
-
+        dims = list(dims)
+        coords = dict(coords)
+        if dims[:2] != default_dims:
+            dims = default_dims + dims
```
[the remark this answers is not in the document]
Comment (ColCarroll) [reply arviz/utils/xarray_utils.py]: yes, it always assumes the first two are chain and draw. You make a good point below that i can clarify the logic. Mostly, I am taking advantage that `np.atleast_3d` works like this: 
```
np.atleast_3d(np.empty(100)).shape == (1, 100, 1)
```
and 
```
np.atleast_3d(np.empty(4, 100)).shape == (4, 100, 1)
```
So if the passed array has shape `(100, )`, I will "squeeze" it at the end and get shape `(1, 100)`, but if the passed array has shape `(1, 100, 1)`, I will respect that and return something with shape `(1, 100, 1)`.

**Pull request merged**